File: ForexConnectAPI-1.1.2-Linux-i686/samples/cpp/SimpleSamples/TableManagerSamples/CreateELS/source/tablelistener.cpp
#include "tablelistener.h"

TableListener::TableListener(void)
{
    mEvent = uni::CreateEvent(0, FALSE, FALSE, 0);
    mOrderID = "";
    mRef = 1;
}

TableListener::~TableListener(void)
{
    uni::CloseHandle(mEvent);
}

long TableListener::addRef()
{
    return InterlockedIncrement(&mRef);
}

long TableListener::release()
{
    InterlockedDecrement(&mRef);
    if(mRef == 0)
        delete this;
    return mRef;
}

HANDLE TableListener::getEvent()
{
    return mEvent;
}

void TableListener::onStatusChanged(O2GTableStatus status)
{
    switch(status)
    {
    case Refreshed:
        uni::SetEvent(mEvent);
        break;
    case Refreshing:
    case Initial:
    case Failed:
        break;
    }
}

void TableListener::onAdded(const char *rowID, IO2GRow *row)
{
    if (row->getTableType() == Orders)
    {
        IO2GOrderRow *orderRow = static_cast<IO2GOrderRow *>(row);
        if (mRequestID == orderRow->getRequestID())
        {
            std::string sOrderType = orderRow->getType();
            std::cout << "Added order. ID: " << orderRow->getOrderID() << \
                " (" << sOrderType << ")" << std::endl;
            mOrderID = orderRow->getOrderID();
            uni::SetEvent(mEvent);
        }
    }
}

void TableListener::onChanged(const char *rowID, IO2GRow *row)
{
    if (row->getTableType() == Orders)
    {
        IO2GOrderRow *orderRow = static_cast<IO2GOrderRow *>(row);
        if (mRequestID == orderRow->getRequestID())
        {
            std::cout << "Changed order. ID: " << orderRow->getOrderID() << std::endl;
            uni::SetEvent(mEvent);
        }
    }
}

void TableListener::onDeleted(const char *rowID, IO2GRow *row)
{
    if (row->getTableType() == Orders)
    {
        IO2GOrderRow *orderRow = static_cast<IO2GOrderRow *>(row);
        if (mRequestID == orderRow->getRequestID())
        {
            std::cout << "Delete order. ID: " << orderRow->getOrderID() << std::endl;
            uni::SetEvent(mEvent);
        }
    }
}

File: ForexConnectAPI-1.1.2-Linux-i686/samples/java/SimpleSamples/TableManagerSamples/OpenPosition/src/TableListener.java
package openposition;

import com.fxcore2.IO2GTableListener;
import com.fxcore2.O2GRow;
import com.fxcore2.O2GTableStatus;
import com.fxcore2.O2GOrderRow;

public class TableListener implements IO2GTableListener {
  
    String mRequestID = null;
    
    public void setRequest(String requestID ) {
        mRequestID = requestID;
    }
    
    // Implementation of IO2GTableListener interface public method onAdded
     public void onAdded(String rowID, O2GRow rowData) {    
        O2GOrderRow orderRow = (O2GOrderRow)(rowData);
        if (mRequestID.equals(orderRow.getRequestID())) {
            System.out.println("Order created. Order ID is " +orderRow.getOrderID());
       } 
    }

    // Implementation of IO2GTableListener interface public method onChanged
    public void onChanged(String rowID, O2GRow rowData) {
    }
    
    // Implementation of IO2GTableListener interface public method onDeleted
    public void onDeleted(String rowID, O2GRow rowData) {
    }

    // Implementation of IO2GTableListener interface public method onStatus
    public void onStatusChanged(O2GTableStatus status) {
        switch(status) {
            case REFRESHED:
                System.out.println("\nYour refresh request was successfull");
                break;
            case REFRESHING:
            case INITIAL:
            case FAILED:
               break;
        }
    }
}
